Optimized ExpectationMaximization

Sand3r- opened the pull request:
Also additional enhancements for Matrix class, and optimized versions of Gaussian Distribution function
and class.

Sand3r- committed:
Implement Optimized version of EM algorithm

Also add optimized versions of Gaussian Distribution function
and class.
Add new constructors for Matrix class.

Sand3r- commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Implement Optimized version of EM algorithm](https://trello.com/c/jYIGnC5r/497-implement-optimized-version-of-em-algorithm)

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/ExpectationRunnerOpt.h:
```diff
@@ -0,0 +1,113 @@
+                    Gaussian(mzs[i], m_Components[lastK].mean,
+                        m_Components[lastK].deviation);
+                m_AffilationMatrix[lastK][i] = probability;
+                m_Sums[i] = 1.0 / (m_Sums[i] + probability);
```
Why does this differ from the version above? Could you add any comment or whatever?

Sand3r- replied:
The formula that is computed in this function is represented by:
![image](https://user-images.githubusercontent.com/4967505/40992964-87ec0c70-68f8-11e8-99e4-630531869e15.png)
Since the divison is a very costly operation, I decided to minimize it in such a way:
1. Compute the sum of probabilities (denominator) for each component from j=0 up to K-1 and save it in sums[i]
2. For the last component, add the final probability and take an inverse of the sums.
3. That way, in the loop below, we no longer have to perform division for each `K*N` elements, but we can apply a much more efficient multiplication. That way, we save ourselves (K-1)*N division operations.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/ExpectationRunnerOpt.h:
```diff
@@ -0,0 +1,113 @@
+    private:
+        SpectrumView m_Spectrum;
+        Matrix<DataType> &m_AffilationMatrix;
+        std::vector<GaussianComponent> &m_Components;
```
Who is the owner of this data? What if the owner gets deleted?

Sand3r- replied:
Owner of the data is the function using the class. I don't like this design either, but I had to implement it this way in order to conform with exisiting API which assumed storing the references to matrix and components upon creation. This will be refactored in the future.

Daroslaw replied:
I think that passing these as arguments into `Expectation` function is a way to go.

gmrukwa replied:
If design is bad, fix it right now. We do not want to merge poorly designed code. What is purpose of CR then? @Daroslaw is right in this case - this should be an argument into `Expectation` function or any other way around, not like this.

You cannot ensure from this point, that owner of this data does not get deleted. Situation you described is just one of possible usages. You do not prevent any other usage.

It's poorly designed, if correctness does not follow from design.

I'd accept it, when there is no other way. But there is.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/Matrix.h:
```diff
@@ -31,29 +31,78 @@ struct Matrix
+    /// Move constructor
+    /// </summary>
+    /// <param name="other">Old matrix object.</param>
+    Matrix(Matrix& other)
```
This isn't move constructor and definitely should not be here. It's better to delete this.

Sand3r- replied:
I have no idea how did this end up here, thanks for catching it.

gmrukwa replied:
I've even meant to make `Matrix(Matrix& other) = delete;` and `Matrix(const Matrix& other) = delete;`. This should prevent us from making implicit copies.

gmrukwa reviewed (commented):
What is the difference in benchmarks?

Sand3r- committed:
Remove copy constructor from Matrix class

Sand3r- commented:
These changes have improved the efficiency of these functions 3 times, however it was the Expectation step that used to take so long.

Daroslaw commented on src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h:
```diff
@@ -19,14 +19,55 @@ limitations under the License.
+            [MIN_WEIGHT](GaussianComponent& component)
+            { return component.weight < MIN_WEIGHT; });
+    size_t newSize = pastEndIter - components.begin();
+    components.resize(newSize);
```
`resize` is better than `erase`?

Sand3r- replied:
I checked the MSVC implementations, and noticed that resize has one less branch.

gmrukwa replied:
We are targeting multiplatform. Deployment will run clang / gcc.

Sand3r- replied:
Kinda hard to tell what's better at clang.
[clang - erase](https://github.com/llvm-mirror/libcxx/blob/master/include/vector#L1708)
[clang - resize](https://github.com/llvm-mirror/libcxx/blob/master/include/vector#L2037)
Still, both functions provide the same effect.

Daroslaw commented on src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h:
```diff
@@ -19,14 +19,55 @@ limitations under the License.
+{
+    DataType sumOfHeightDifferences = 0.0;
+    DataType sumOfScaledVarianceDifferences = 0.0;
+    DataType justASum = 0.0;
```
xD

Sand3r- replied:
Oh what xDDD cannot believe that I've actually left it that way xD

Daroslaw commented on src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h:
```diff
@@ -19,14 +19,55 @@ limitations under the License.
+        sumOfScaledVarianceDifferences +=
+            abs(newModel[i].deviation * newModel[i].deviation -
+                oldModel[i].deviation * oldModel[i].deviation) /
+               (newModel[i].deviation * newModel[i].deviation);
```
I think you can squish some more performance by writing it as:
`1 - (oldModel[i].deviation * oldModel[i].deviation) / (newModel[i].deviation * newModel[i].deviation)`, because:
![image](https://user-images.githubusercontent.com/7458086/41052299-b324e32e-69b8-11e8-9283-66dc943019e3.png)
Or at least buffer the `newModel[i].deviation * newModel[i].deviation`.

Sand3r- replied:
As we've discuessed on PM, that is not so easy to implement due to abs(),

Daroslaw replied:
😢

Sand3r- committed:
Ensure nonzero probabilities in E step

Daroslaw commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerOpt.h:
```diff
@@ -0,0 +1,125 @@
+        /// <summary>
+        /// Empty function made to conform with interface.
+        /// </summary>
+        void UpdateStdDeviations()
```
Are these functions meant to be filled in the future? If they are only to conform with interface why are they even called anywhere?

Sand3r- replied:
Due to existing API this cannot be changed without refining everything else. 
To be refactored in the future.

gmrukwa replied:
Such approach is bad. If API is unsuitable, make it fit. It's better to do it now, than to carry it longer, to let problem grow.

Daroslaw commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerOpt.h:
```diff
@@ -0,0 +1,125 @@
+    private:
+        SpectrumView m_Spectrum;
+        Matrix<DataType> &m_AffilationMatrix;
+        std::vector<GaussianComponent> &m_Components;
```
Same as before

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/GaussianDistribution.h:
```diff
@@ -28,8 +28,48 @@ namespace spectre::unsupervised::gmm
+    }
+
+private:
+    DataType m_Mean;
```
Aren't these `const`?

Sand3r- committed:
Redesign API of ExpectationMaximization

Previously the EM-related classes defined the data they would
operate on during constructions, and functions performing the
calculations where parameterless and void.

Now the constructors are more vague and treat parameters as a
way to initialize their internal cache-helpers, while functions
accept all the data they work on as parameters.

Sand3r- committed:
Ensure constness of GaussianDistribution params

Also extract the computation of inverse square root of 2pi to
a literal constant, as to not compute it during the runtime.

Daroslaw commented on src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h:
```diff
@@ -74,77 +74,69 @@ static inline DataType CalculateDelta(GaussianMixtureModel& oldModel,
         } while (delta > epsilon);
 
-        return m_Components; // Consider sorting before return
+        return components;
```
If you modify the argument you don't need to return it. Entire function may be of type `void`

Sand3r- replied:
Yes I am aware of it, but to comply with the unwritten standard of native-algorithms libs, I decided to return it anyway, because otherwise @gmrukwa would complain.

Sand3r- committed:
Update appveyor version to 4.7.0

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h:
```diff
@@ -19,103 +19,125 @@ limitations under the License.
 #include <random>
 #include "Spectre.libException/NullPointerException.h"
+#include "Spectre.libStatistics/Math.h"
 #include "DataTypes.h"
```
I know this is not a part of this review, but please create task to make these paths w.r.t. solution root. Otherwise NuGet package won't work.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/ExpectationRunnerRef.h:
```diff
@@ -40,60 +40,60 @@ limitations under the License.
+    /// fetched.</param>
+    ExpectationRunnerRef(SpectrumView spectrum)
     {
+        spectrum.mzs.size(); // Needed to avoid warning of unused param.
```
It was usually solved by definition like `ExpectationRunnerRef(SpectrumView /*spectrum*/)`. This avoids strange, unnecessary call.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerRef.h:
```diff
@@ -48,81 +48,114 @@ limitations under the License.
         // from the document
         DataType totalDataSize = 0.0;
-        for (unsigned i = 0; i < (unsigned)m_Spectrum.mzs.size(); i++)
+        for (unsigned i = 0; i < (unsigned)spectrum.mzs.size(); i++)
```
Wait... Are you computing this each time you do _Maximization_ step? This should never change. What about precomputing this once and not repeating this operation?

Sand3r- replied:
It is precomputed in the optimized version MaximizationRunnerOpt.h

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerRef.h:
```diff
@@ -48,81 +48,114 @@ limitations under the License.
-        : m_Spectrum(spectrum), m_AffilationMatrix(affilationMatrix), m_Components(components)
+    void Maximization(Matrix<DataType> &affilationMatrix,
+        SpectrumView spectrum, GaussianMixtureModel& components)
     {
```
Why do we have definitions in `.h` file?

Sand3r- replied:
That's how the previous API used to look like. It limits the number of translation units and allows compiler for doing better optimizations.

gmrukwa replied:
This requires recompilation in each dependent project, while step itself will be used in one context: context of GMM. I'd understand such approach with your argument for GMM itself, but not for its elements. Maybe you could clarify it a bit more?

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerRef.h:
```diff
@@ -48,81 +48,114 @@ limitations under the License.
+            for (unsigned i = 0; i < (unsigned)spectrum.mzs.size(); i++)
             {
-                weight += m_AffilationMatrix[i][k] * m_Spectrum.intensities[i];
+                weight += affilationMatrix[k][i] * spectrum.intensities[i];
```
Here is probably great underflow. You are summing like 200 000 floats in real data. Really small floats, coming from gaussian distribution. Check [this thread](https://stackoverflow.com/questions/2148149/how-to-sum-a-large-number-of-float-number).

Sand3r- replied:
This is a reference version which won't be likely to be used.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerRef.h:
```diff
@@ -48,81 +48,114 @@ limitations under the License.
             {
-                denominator += m_AffilationMatrix[i][k] * m_Spectrum.intensities[i];
-                numerator += m_AffilationMatrix[i][k] * m_Spectrum.mzs[i] * m_Spectrum.intensities[i];
+                denominator += affilationMatrix[k][i] * spectrum.intensities[i];
```
Possible underflows.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerRef.h:
```diff
@@ -133,24 +166,20 @@ class MaximizationRunnerRef
-                denominator += m_AffilationMatrix[i][k] * m_Spectrum.intensities[i];
-                numerator += m_AffilationMatrix[i][k] *
-                    pow(m_Spectrum.mzs[i] - m_Components[k].mean, 2) * m_Spectrum.intensities[i];
+                denominator += affilationMatrix[k][i] * spectrum.intensities[i];
```
Possible underflows.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/Matrix.h:
```diff
@@ -31,29 +31,68 @@ struct Matrix
+        memset(data[0], 0, length * sizeof(Type));
     }
 
+    /// <summary>
```
If this class gets extended even a bit more, switch to `cv::Mat` or just consider using Eigen.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/Matrix.h:
```diff
@@ -66,6 +105,22 @@ struct Matrix
+    {
+        for (unsigned i = 0; i < length; i++)
+        {
+            if (data[0][i] != other.data[0][i])
```
We are using this for `float`s mostly. What about epsilon-based comparison?

Sand3r- replied:
That was made just for the purpose of tests with indices (yet to come in Dyn Prog initialization)). I'd need to create a template specialized method for it.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/GaussianDistribution.h:
```diff
@@ -28,8 +28,47 @@ namespace spectre::unsupervised::gmm
+    /// </summary>
+    /// <param name="mean">Mean of the distribution.</param>
+    /// <param name="std">Standard deviation of the distribution.</param>
+    GaussianDistribution(DataType mean, DataType std) : m_Mean(mean),
```
We will need information on `std` and _weight_ in final model - this is used for components merging, etc. How do you preserve this? Do you return this at some point other way?

Sand3r- replied:
These are saved in Gaussian Components.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerOpt.h:
```diff
@@ -0,0 +1,106 @@
+                DataType meanNumerator = 0.0;
+                for (unsigned i = 0; i < (unsigned)mzs.size(); i++)
+                {
+                    probabilitySum += affilationMatrix[k][i];
```
Possible underflows.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/MaximizationRunnerOpt.h:
```diff
@@ -0,0 +1,106 @@
+                for (unsigned i = 0; i < (unsigned)mzs.size(); i++)
+                {
+                    DataType diff = mzs[i] - components[k].mean;
+                    stdNumerator += affilationMatrix[k][i] * diff * diff;
```
Possible underflows.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/ExpectationRunnerOpt.h:
```diff
@@ -0,0 +1,105 @@
+                    Gaussian(mzs[i], components[k].mean,
+                        components[k].deviation);
+                affilationMatrix[k][i] = probability;
+                m_Sums[i] += probability;
```
Possible underflows.

gmrukwa reviewed (commented):
Please explain, how did you approach underflow problem in the EM for real data we will be processing (200k m/z-s). When this gets clear, I click *Approve*.

Sand3r- committed:
Introduce Kahan summation algorithm for EM

Introducing Kahan summation [1] to improve accuracy of Expecatation
and Maximization Optimized steps. The scheme allows for improved
error correction of floating point addition of numbers of highly dispersed
magnitudes.

[1] https://en.wikipedia.org/wiki/Kahan_summation_algorithm

Sand3r- committed:
Change relative to absolute paths in GMM project

Since NuGet needs absolute paths to create packages, all
existing paths to header files have been changed to be
relative from the point of view of the Solution (= absolute)

Sand3r- committed:
Remove empty file

Sand3r- committed:
Change construction signatures of EM to avoid warning

Parameter names have been removed to avoid warning of unsused
variables.

gmrukwa commented on src/Spectre.libGaussianMixtureModelling/ExpectationRunnerOpt.h:
```diff
@@ -73,13 +73,19 @@ class ExpectationRunnerOpt
         const unsigned numberOfComponents = (unsigned)components.size();
         for (unsigned k = 0; k < numberOfComponents; k++)
         {
+            // Using Kahan summation algorithm for improved accuracy
```
Can we extract this into a reusable element?

Sand3r- replied:
I've considered doing that, but in another PR.

gmrukwa replied:
Add a task then.

gmrukwa reviewed (approved):
Seems OK now :) Good job!
